
<h1 align="center"> Battle of the Neighbourhoods - Toronto </h1>

Author: Thomas George Thomas

This notebook contains Questions 1, 2 & 3 of the Assignment. They have been segregated by Section headers


In [1]:
import pandas as pd

# Question 1

## Importing Data

In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we are able to make the connection to the page

In [4]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code  ...                                      Neighbourhood
 0           M1A  ...                                       Not assigned
 1           M2A  ...                                       Not assigned
 2           M3A  ...                                          Parkwoods
 3           M4A  ...                                   Victoria Village
 4           M5A  ...                          Regent Park, Harbourfront
 ..          ...  ...                                                ...
 175         M5Z  ...                                       Not assigned
 176         M6Z  ...                                       Not assigned
 177         M7Z  ...                                       Not assigned
 178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...
 179         M9Z  ...                                       Not assigned
 
 [180 rows x 3 columns],
                                                   0   ...   17
 0                               

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

We need the first table alone, so dropping the other tables

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code  ...                                      Neighbourhood
0           M1A  ...                                       Not assigned
1           M2A  ...                                       Not assigned
2           M3A  ...                                          Parkwoods
3           M4A  ...                                   Victoria Village
4           M5A  ...                          Regent Park, Harbourfront
..          ...  ...                                                ...
175         M5Z  ...                                       Not assigned
176         M6Z  ...                                       Not assigned
177         M7Z  ...                                       Not assigned
178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...
179         M9Z  ...                                       Not assigned

[180 rows x 3 columns]

Dropping Borough which are not assigned

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code  ...                                      Neighbourhood
2           M3A  ...                                          Parkwoods
3           M4A  ...                                   Victoria Village
4           M5A  ...                          Regent Park, Harbourfront
5           M6A  ...                   Lawrence Manor, Lawrence Heights
6           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
160         M8X  ...      The Kingsway, Montgomery Road, Old Mill North
165         M4Y  ...                               Church and Wellesley
168         M7Y  ...  Business reply mail Processing Centre, South C...
169         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [8]:
df = df.groupby(['Postal Code']).head()
df

Postal Code  ...                                      Neighbourhood
2           M3A  ...                                          Parkwoods
3           M4A  ...                                   Victoria Village
4           M5A  ...                          Regent Park, Harbourfront
5           M6A  ...                   Lawrence Manor, Lawrence Heights
6           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
160         M8X  ...      The Kingsway, Montgomery Road, Old Mill North
165         M4Y  ...                               Church and Wellesley
168         M7Y  ...  Business reply mail Processing Centre, South C...
169         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [10]:
df = df.reset_index()
df

index  ...                                      Neighbourhood
0        2  ...                                          Parkwoods
1        3  ...                                   Victoria Village
2        4  ...                          Regent Park, Harbourfront
3        5  ...                   Lawrence Manor, Lawrence Heights
4        6  ...        Queen's Park, Ontario Provincial Government
..     ...  ...                                                ...
98     160  ...      The Kingsway, Montgomery Road, Old Mill North
99     165  ...                               Church and Wellesley
100    168  ...  Business reply mail Processing Centre, South C...
101    169  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
102    178  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 4 columns]

In [11]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code  ...                                      Neighbourhood
0           M3A  ...                                          Parkwoods
1           M4A  ...                                   Victoria Village
2           M5A  ...                          Regent Park, Harbourfront
3           M6A  ...                   Lawrence Manor, Lawrence Heights
4           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
98          M8X  ...      The Kingsway, Montgomery Road, Old Mill North
99          M4Y  ...                               Church and Wellesley
100         M7Y  ...  Business reply mail Processing Centre, South C...
101         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
102         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

In [12]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

# Question 2

Installing geocoder

In [13]:
pip install geocoder

     |████████████████████████████████| 102kB 2.2MB/s 


In [14]:
import geocoder # import geocoder

Tried the below approach, ran for 20 mins, then killed it. Changing the code cell to Text for now so that the run all execution doesn't stop.

```python
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
```

Alternatively, as suggested in the assignment, Importing the CSV file from the URL

In [15]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [16]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [17]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [18]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [19]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  ...   Latitude  Longitude
0           M3A        North York  ...  43.753259 -79.329656
1           M4A        North York  ...  43.725882 -79.315572
2           M5A  Downtown Toronto  ...  43.654260 -79.360636
3           M6A        North York  ...  43.718518 -79.464763
4           M7A  Downtown Toronto  ...  43.662301 -79.389494
..          ...               ...  ...        ...        ...
98          M8X         Etobicoke  ...  43.653654 -79.506944
99          M4Y  Downtown Toronto  ...  43.665860 -79.383160
100         M7Y      East Toronto  ...  43.662744 -79.321558
101         M8Y         Etobicoke  ...  43.636258 -79.498509
102         M8Z         Etobicoke  ...  43.628841 -79.520999

[103 rows x 5 columns]

In [20]:
combined_data.shape

(103, 5)

**Solution:** We get 103 rows as expected when we do a inner join, so we have good data.

# Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [21]:
from geopy.geocoders import Nominatim 

In [22]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [23]:
import folium

In [60]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [28]:
CLIENT_ID = 'LDIJF4KI5VGMMA3NNDLFZWHR12TCMNTUL0TUC3QPZ3SJD040' 
CLIENT_SECRET = 'EOOOZ3EF5N0FOMNUJVTDV0SXVUVVEBMWPFXMNBK1R5K4H55A'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: LDIJF4KI5VGMMA3NNDLFZWHR12TCMNTUL0TUC3QPZ3SJD040
CLIENT_SECRET:EOOOZ3EF5N0FOMNUJVTDV0SXVUVVEBMWPFXMNBK1R5K4H55A


Next, we create a function to get all the venue categories in Toronto

In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [30]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [31]:
venues_in_toronto.shape

(1317, 5)

So we have 1317 records and 5 columns. Checking sample data


In [32]:
venues_in_toronto.head()

Neighbourhood  ...         Venue Category
0         Parkwoods  ...                   Park
1         Parkwoods  ...      Food & Drink Shop
2  Victoria Village  ...           Hockey Arena
3  Victoria Village  ...            Coffee Shop
4  Victoria Village  ...  Portuguese Restaurant

[5 rows x 5 columns]

Checking the Venues based on Neighbourhood

In [33]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  ...         Venue Category
0                                             Parkwoods  ...                   Park
1                                             Parkwoods  ...      Food & Drink Shop
2                                      Victoria Village  ...           Hockey Arena
3                                      Victoria Village  ...            Coffee Shop
4                                      Victoria Village  ...  Portuguese Restaurant
...                                                 ...  ...                    ...
1302  Mimico NW, The Queensway West, South of Bloor,...  ...            Wings Joint
1303  Mimico NW, The Queensway West, South of Bloor,...  ...           Burger Joint
1304  Mimico NW, The Queensway West, South of Bloor,...  ...         Discount Store
1305  Mimico NW, The Queensway West, South of Bloor,...  ...        Supplement Shop
1306  Mimico NW, The Queensway West, South of Bloor,...  ...                 Bakery

[405 rows x 5 columns]

So there are 405 records for each neighbourhood.

Checking for the maximum venue categories

In [34]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  ...                            Venue
Venue Category                                                         ...                                 
Accessories Store                    Lawrence Manor, Lawrence Heights  ...              Ardene Shoes Outlet
Airport                                                     Downsview  ...  Toronto Downsview Airport (YZD)
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...  ...                Billy Bishop Café
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...  ...                           Gate 8
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...  ...                    Porter Lounge
...                                                               ...  ...                              ...
Warehouse Store                                      Thorncliffe Park  ...                           Costco
Wine Bar                                              Studio District  ...                  Paris Paris Bar
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...  ...                       Wingporium
Women's Store                                     Caledonia-Fairbanks  ...                    Maximum Woman
Yoga Studio                            University of Toronto, Harbord  ...                        YogaSpace

[232 rows x 4 columns]

There are around 232 different types of Venue Categories. Interesting!

## One Hot encoding the venue Categories

In [35]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  ...  Women's Store  Yoga Studio
0                     0        0  ...              0            0
1                     0        0  ...              0            0
2                     0        0  ...              0            0
3                     0        0  ...              0            0
4                     0        0  ...              0            0
...                 ...      ...  ...            ...          ...
1312                  0        0  ...              0            0
1313                  0        0  ...              0            0
1314                  0        0  ...              0            0
1315                  0        0  ...              0            0
1316                  0        0  ...              0            0

[1317 rows x 232 columns]

Adding the neighbourhood to the encoded dataframe

In [36]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  ...  Women's Store  Yoga Studio
0         Parkwoods                  0  ...              0            0
1         Parkwoods                  0  ...              0            0
2  Victoria Village                  0  ...              0            0
3  Victoria Village                  0  ...              0            0
4  Victoria Village                  0  ...              0            0

[5 rows x 233 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood 

In [37]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  ...  Yoga Studio
0                                        Agincourt  ...          0.0
1                           Alderwood, Long Branch  ...          0.0
2  Bathurst Manor, Wilson Heights, Downsview North  ...          0.0
3                                  Bayview Village  ...          0.0
4                Bedford Park, Lawrence Manor East  ...          0.0

[5 rows x 233 columns]

Let's make a function to get the top most common venue categories

In [38]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [39]:
import numpy as np

There are way too many venue categories, we can take the top 10 to cluster the neighbourhoods

In [40]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood  ... 10th Most Common Venue
0                                        Agincourt  ...         Discount Store
1                           Alderwood, Long Branch  ...           Dessert Shop
2  Bathurst Manor, Wilson Heights, Downsview North  ...            Pizza Place
3                                  Bayview Village  ...    Distribution Center
4                Bedford Park, Lawrence Manor East  ...                Butcher

[5 rows x 11 columns]

Let's make the model to cluster our Neighbourhoods

In [41]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [42]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

Checking the labelling of our model

In [43]:
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 3, 2, 0, 0, 0,
       2, 0, 2, 0, 0, 2, 1, 0, 0, 0, 2, 0, 0, 0, 3, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4,
       0, 1, 0, 0, 0, 0, 2], dtype=int32)

Let's add the clustering Label column to the top 10 common venue categories

In [44]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [45]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough  ... 9th Most Common Venue  10th Most Common Venue
0         M3A        North York  ...        Discount Store                   Diner
1         M4A        North York  ...          Dessert Shop                   Diner
2         M5A  Downtown Toronto  ...        Chocolate Shop             Yoga Studio
3         M6A        North York  ...      Cuban Restaurant                 Dog Run
4         M7A  Downtown Toronto  ...    Persian Restaurant      College Auditorium

[5 rows x 16 columns]

Drop all the NaN values to prevent data skew

In [46]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [47]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [73]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [49]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  ...  10th Most Common Venue
1          North York  ...                   Diner
2    Downtown Toronto  ...             Yoga Studio
3          North York  ...                 Dog Run
4    Downtown Toronto  ...      College Auditorium
7          North York  ...    Caribbean Restaurant
..                ...  ...                     ...
97   Downtown Toronto  ...    Gym / Fitness Center
98          Etobicoke  ...      Dim Sum Restaurant
99   Downtown Toronto  ...              Restaurant
100      East Toronto  ...           Garden Center
102         Etobicoke  ...         Supplement Shop

[82 rows x 12 columns]

Cluster 2

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  ...  10th Most Common Venue
0         North York  ...                   Diner
21              York  ...                   Diner
32       Scarborough  ...                   Diner
52        North York  ...                   Diner
85       Scarborough  ...                   Diner
91  Downtown Toronto  ...                   Diner

[6 rows x 12 columns]

Cluster 3

In [51]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  ...  10th Most Common Venue
6       Scarborough  ...                   Diner
16             York  ...          Discount Store
35        East York  ...                 Dog Run
44      Scarborough  ...              Donut Shop
49       North York  ...                 Dog Run
61  Central Toronto  ...                 Dog Run
66       North York  ...          Discount Store
77        Etobicoke  ...                   Diner

[8 rows x 12 columns]

Cluster 4

In [52]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
57   North York             3.0  ...   Distribution Center         Discount Store
101   Etobicoke             3.0  ...   Distribution Center         Discount Store

[2 rows x 12 columns]

Cluster 5

In [53]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels  ... 9th Most Common Venue 10th Most Common Venue
64    York             4.0  ...   Distribution Center         Discount Store

[1 rows x 12 columns]

We have successfully cluster Toronto neighbourhood based on venue categories!